lib/llvmopencl/UnreachablesToReturns.cc

```diff
@@ -138,106 +150,164 @@ static bool convertUnreachablesToReturns(Function &F) {
+        LLVM_DEBUG(
+            dbgs()
+            << "A switch case with only a default which goes to unreachable\n");
+        NewUnreachableBBs.insert(PredSwitch->getParent());
+      }
+    }
+    LLVM_DEBUG(dbgs() << "switch case modified:\n");
+    LLVM_DEBUG(PredSwitch->getParent()->dump());
+  }
 }
 
 static bool deleteBlocksWithUnreachable(Function &F) {
 
   SmallBBSet UnreachableBBs;
 
-  bool Changed = removeUnreachableSwitchCases(F);
-
   for (Function::iterator I = F.begin(), E = F.end(); I != E; ++I) {
     BasicBlock &BB = *I;
     assert(BB.getTerminator());
     if (isa<UnreachableInst>(BB.getTerminator())) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "UNREACHABLE found: deleting BB in "
-                        << F.getName().str() << "\n");
-#endif
+      LLVM_DEBUG(dbgs() << "UNREACHABLE found, deleting BB\n");
+      LLVM_DEBUG(BB.dump());
       UnreachableBBs.insert(&BB);
     }
   }
 
   if (UnreachableBBs.empty())
-    return Changed;
+    return false;
 
-  // check BB predecessors recursively, and disconnect them
-  // from blocks which contain an unreachable
-  SmallBBSet UnreachableBBs2;
-  for (auto *BB : UnreachableBBs) {
-    detachBBFromPredecessor(BB, UnreachableBBs2);
+  // Check BB predecessors recursively, and disconnect them
+  // from blocks which contain an unreachable.
+  SmallBBSet HandledUnreachableBBs;
+  while (!UnreachableBBs.empty()) {
+    BasicBlock *BB = *UnreachableBBs.begin();
+    detachBBFromPredecessor(*BB, UnreachableBBs);
+    UnreachableBBs.erase(BB);
+    HandledUnreachableBBs.insert(BB);
   }
 
-  // all relevant blocks should now be disconnected (have 0 predecessors)
-  // delete them
-  for (auto *BB : UnreachableBBs) {
-    if (BB->hasNPredecessors(0)) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "deleting BB: \n");
-      BB->dump();
-#endif
-      BB->eraseFromParent();
-    }
-  }
+  while (!HandledUnreachableBBs.empty()) {
 
-  for (auto *BB : UnreachableBBs2) {
-    if (BB->hasNPredecessors(0)) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "deleting BB: \n";
-      BB->dump();
-#endif
-      BB->eraseFromParent();
+    auto CandidateBB = HandledUnreachableBBs.begin();
+
+    // We have to delete the "chains" bottom up to avoid having basic blocks
+    // that refer to the values produced by the predecessors in the stem.
+    while (!isa<UnreachableInst>((*CandidateBB)->getTerminator()))
+      ++CandidateBB;
+    assert(CandidateBB != HandledUnreachableBBs.end());
+    BasicBlock *BB = *CandidateBB;
+
+    LLVM_DEBUG(dbgs() << "Deleting BB: \n");
+    LLVM_DEBUG(BB->dump());
+
+    // The deleted BB can have multiple predecessors.
+    SmallMapVector<Instruction *, Instruction *, 8> Replacements;
+    for (BasicBlock *Pred : predecessors(BB))
+      Replacements.insert(std::make_pair(
+          Pred->getTerminator(), new UnreachableInst(Pred->getContext())));
```

Comment:
FWIW, if this were done in the `detachBBFromPredecessor()`, the outer loop probably would simplify to:
```
for (auto *BB: HandledUnreachableBBs)
  BB->eraseFromParent();
```